File: desktop/ShareCLITray/Sources/ShareCLICore/OperatorDisplay.swift
/// OperatorDisplay.swift — compact gate / host_watch tray strings + thermal visuals
/// (FR-007 / AC-007.56 text, AC-007.57 icon/badge/color).
///
/// Parity with `sharecli-tray-linux/src/operator_display.rs` golden strings.

import Foundation
import SwiftUI

public enum TrayGateSeverity: String, Equatable {
    case normal
    case warning
    case critical
    case offline
}

public struct TrayGateVisual: Equatable {
    public let severity: TrayGateSeverity
    public let decisionClass: String
    public let thermalClass: String
    public let colorHex: String
    public let badgeLabel: String
    public let swiftSymbolName: String

    public var swiftColor: Color {
        switch severity {
        case .normal: return Color(red: 0.247, green: 0.725, blue: 0.314) // #3fb950
        case .warning: return Color(red: 0.824, green: 0.600, blue: 0.133) // #d29922
        case .critical: return Color(red: 0.973, green: 0.318, blue: 0.286) // #f85149
        case .offline: return Color(red: 0.824, green: 0.600, blue: 0.133)
        }
    }
}

public enum OperatorDisplay {
    public static func formatBytesCompact(_ n: UInt64) -> String {
        if n >= 1_048_576 {
            return String(format: "%.1f MB", Double(n) / 1_048_576.0)
        }
        if n >= 1024 {
            return String(format: "%.1f KB", Double(n) / 1024.0)
        }
        return "\(n) B"
    }

    public static func formatGateTrayLine(_ gate: GateStatusSnapshot) -> String {
        "Gate [\(gate.gate_decision)] · \(gate.thermal_pressure) · agents \(gate.detected_agents) · \(gate.agent_contention)"
    }

    public static func formatGateRssTrayLine(_ gate: GateStatusSnapshot) -> String {
        "Agent RSS: \(formatBytesCompact(gate.agent_total_rss_bytes))"
    }

    public static func formatHostWatchTrayLine(_ host: HostResourceWatchJson) -> String {
        String(
            format: "Host load %.2f · FDs %@ · RSS %@",
            host.load_1m,
            String(host.fd_count),
            formatBytesCompact(host.mem_rss_bytes)
        )
    }

    public static func formatHostNetTrayLine(_ host: HostResourceWatchJson) -> String {
        "Net RX \(formatBytesCompact(host.net_rx_bytes)) · TX \(formatBytesCompact(host.net_tx_bytes))"
    }

    public static func formatOperatorTrayLines(
        gate: GateStatusSnapshot,
        host: HostResourceWatchJson
    ) -> [String] {
        [
            formatGateTrayLine(gate),
            formatGateRssTrayLine(gate),
            formatHostWatchTrayLine(host),
            formatHostNetTrayLine(host),
        ]
    }

    public static func formatOperatorStatusSummary(
        gate: GateStatusSnapshot,
        host: HostResourceWatchJson
    ) -> String {
        [
            formatGateTrayLine(gate),
            formatGateRssTrayLine(gate),
            formatHostWatchTrayLine(host),
        ].joined(separator: " | ")
    }

    /// Compact menu bar title for NSStatusItem (AC-007.62).
    public static func formatMenuBarTitleLine(visual: TrayGateVisual, health: HealthSnapshot) -> String {
        " \(visual.badgeLabel) · \(health.managed_processes) | \(health.used_memory_mb)M"
    }

    /// Offline menu bar title when monitoring.report is unavailable (AC-007.62).
    public static func formatMenuBarTitleOfflineLine(visual: TrayGateVisual) -> String {
        " \(visual.badgeLabel) · offline"
    }

    /// Pool capacity line from `pool.status` IPC (AC-007.69).
    public static func formatPoolTrayLine(_ pool: PoolSnapshot) -> String {
        let healthLabel = pool.healthy ? "healthy" : "degraded"
        return "Pool node \(pool.node_total)/\(pool.node_idle) idle · bun \(pool.bun_total)/\(pool.bun_idle) idle · max \(pool.max_per_type) · \(healthLabel)"
    }

    /// Proc-scan summary line from `status.snapshot` IPC (AC-007.69).
    public static func formatStatusSnapshotTrayLine(_ status: StatusSnapshot) -> String {
        "Proc scan \(status.scanned) · watched \(status.watched) · \(status.total_processes) managed · \(status.agents.count) agent row(s)"
    }

    /// Supplementary pool + status operator lines (AC-007.69).
    public static func formatPoolStatusOperatorLines(
        pool: PoolSnapshot,
        status: StatusSnapshot
    ) -> [String] {
        [formatPoolTrayLine(pool), formatStatusSnapshotTrayLine(status)]
    }

    public static func resolveGateDecisionClass(_ gateDecision: String) -> String {
        switch gateDecision {
        case "ADMIT": return "gate-admit"
        case "DENY": return "gate-deny"
        default: return "gate-unavailable"
        }
    }

    public static func resolveThermalClass(_ thermalPressure: String) -> String {
        switch thermalPressure {
        case "GREEN": return ""
        case "YELLOW": return "warning"
        case "RED": return "critical"
        default: return "warning"
        }
    }

    public static func resolveTrayGateVisual(
        thermalPressure: String,
        gateDecision: String,
        connected: Bool
    ) -> TrayGateVisual {
        if !connected {
            return TrayGateVisual(
                severity: .offline,
                decisionClass: "gate-unavailable",
                thermalClass: "warning",
                colorHex: "#d29922",
                badgeLabel: "Offline",
                swiftSymbolName: "wifi.slash"
            )
        }

        let decisionClass = resolveGateDecisionClass(gateDecision)
        let thermalClass = resolveThermalClass(thermalPressure)

        let severity: TrayGateSeverity
        if gateDecision == "DENY" || thermalPressure == "RED" {
            severity = .critical
        } else if gateDecision == "THROTTLE" || thermalPressure == "YELLOW"
            || thermalPressure == "UNAVAILABLE" || decisionClass == "gate-unavailable"
        {
            severity = .warning
        } else {
            severity = .normal
        }

        switch severity {
        case .critical:
            return TrayGateVisual(
                severity: .critical,
                decisionClass: decisionClass,
                thermalClass: thermalClass,
                colorHex: "#f85149",
                badgeLabel: "Critical",
                swiftSymbolName: "flame.fill"
            )
        case .warning:
            let label = thermalPressure == "UNAVAILABLE" ? "Unavailable" : "Warning"
            return TrayGateVisual(
                severity: .warning,
                decisionClass: decisionClass,
                thermalClass: thermalClass,
                colorHex: "#d29922",
                badgeLabel: label,
                swiftSymbolName: "exclamationmark.triangle.fill"
            )
        case .normal:
            return TrayGateVisual(
                severity: .normal,
                decisionClass: decisionClass,
                thermalClass: thermalClass,
                colorHex: "#3fb950",
                badgeLabel: "Normal",
                swiftSymbolName: "cpu"
            )
        case .offline:
            fatalError("unreachable")
        }
    }

    public static func resolveTrayGateVisual(gate: GateStatusSnapshot, connected: Bool) -> TrayGateVisual {
        resolveTrayGateVisual(
            thermalPressure: gate.thermal_pressure,
            gateDecision: gate.gate_decision,
            connected: connected
        )
    }
}

File: crates/sharecli-tray-linux/src/operator_display.rs
//! Compact gate / host_watch strings + thermal gate visuals for tray operator UI
//! (FR-007 / AC-007.56 text, AC-007.57 icon/badge/color).
//!
//! Mirrors dashboard operator panels and proc/status text sections; testable without GUI.

use crate::ipc::{
    GateStatusSnapshot, HealthSnapshot, HostResourceWatchJson, PoolSnapshot, StatusSnapshot,
};

/// Tray thermal/gate severity bucket (dashboard `#thermal-status` + gate decision).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TrayGateSeverity {
    Normal,
    Warning,
    Critical,
    Offline,
}

/// Visual tokens for tray icon / badge / color (parity with `src/dashboard.html`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TrayGateVisual {
    pub severity: TrayGateSeverity,
    /// Dashboard gate CSS class: `gate-admit` | `gate-deny` | `gate-unavailable`.
    pub decision_class: &'static str,
    /// Dashboard thermal CSS suffix: `""` | `warning` | `critical`.
    pub thermal_class: &'static str,
    /// Dashboard design-token hex (`--bb2-pulse-green`, deny red, warm amber).
    pub color_hex: &'static str,
    /// Compact badge label (`Normal`, `Warning`, `Critical`, `Offline`, `Unavailable`).
    pub badge_label: &'static str,
    /// freedesktop icon name for Linux StatusNotifierItem.
    pub linux_icon_name: &'static str,
    /// SF Symbol name for macOS menu bar (Swift parity).
    pub swift_symbol_name: &'static str,
}

/// Map `gate_decision` → dashboard gate CSS class (AC-007.41 / AC-007.57).
pub fn resolve_gate_decision_class(gate_decision: &str) -> &'static str {
    match gate_decision {
        "ADMIT" => "gate-admit",
        "DENY" => "gate-deny",
        _ => "gate-unavailable",
    }
}

/// Map `thermal_pressure` → dashboard `#thermal-status` CSS suffix.
pub fn resolve_thermal_class(thermal_pressure: &str) -> &'static str {
    match thermal_pressure {
        "GREEN" => "",
        "YELLOW" => "warning",
        "RED" => "critical",
        _ => "warning",
    }
}

/// Resolve tray icon/badge/color from live gate fields (AC-007.57).
pub fn resolve_tray_gate_visual(
    thermal_pressure: &str,
    gate_decision: &str,
    connected: bool,
) -> TrayGateVisual {
    if !connected {
        return TrayGateVisual {
            severity: TrayGateSeverity::Offline,
            decision_class: "gate-unavailable",
            thermal_class: "warning",
            color_hex: "#d29922",
            badge_label: "Offline",
            linux_icon_name: "network-offline",
            swift_symbol_name: "wifi.slash",
        };
    }

    let decision_class = resolve_gate_decision_class(gate_decision);
    let thermal_class = resolve_thermal_class(thermal_pressure);

    let severity = if gate_decision == "DENY" || thermal_pressure == "RED" {
        TrayGateSeverity::Critical
    } else if gate_decision == "THROTTLE"
        || thermal_pressure == "YELLOW"
        || thermal_pressure == "UNAVAILABLE"
        || decision_class == "gate-unavailable"
    {
        TrayGateSeverity::Warning
    } else {
        TrayGateSeverity::Normal
    };

    let (color_hex, badge_label, linux_icon_name, swift_symbol_name) = match severity {
        TrayGateSeverity::Critical => ("#f85149", "Critical", "dialog-error", "flame.fill"),
        TrayGateSeverity::Warning => {
            let label = if thermal_pressure == "UNAVAILABLE" { "Unavailable" } else { "Warning" };
            ("#d29922", label, "dialog-warning", "exclamationmark.triangle.fill")
        }
        TrayGateSeverity::Normal => ("#3fb950", "Normal", "utilities-system-monitor", "cpu"),
        TrayGateSeverity::Offline => unreachable!("handled above"),
    };

    TrayGateVisual {
        severity,
        decision_class,
        thermal_class,
        color_hex,
        badge_label,
        linux_icon_name,
        swift_symbol_name,
    }
}

/// Convenience: derive visuals from a gate snapshot + IPC connectivity.
pub fn resolve_tray_gate_visual_from_gate(
    gate: &GateStatusSnapshot,
    connected: bool,
) -> TrayGateVisual {
    resolve_tray_gate_visual(&gate.thermal_pressure, &gate.gate_decision, connected)
}

impl Default for TrayGateVisual {
    fn default() -> Self {
        resolve_tray_gate_visual("UNAVAILABLE", "UNAVAILABLE", false)
    }
}

/// Human-readable byte count for tray lines (parity with dashboard `formatBytes`).
pub fn format_bytes_compact(n: u64) -> String {
    if n >= 1_048_576 {
        format!("{:.1} MB", n as f64 / 1_048_576.0)
    } else if n >= 1024 {
        format!("{:.1} KB", n as f64 / 1024.0)
    } else {
        format!("{n} B")
    }
}

/// Primary thermal gate line for tray menus / status bars.
pub fn format_gate_tray_line(gate: &GateStatusSnapshot) -> String {
    format!(
        "Gate [{}] · {} · agents {} · {}",
        gate.gate_decision, gate.thermal_pressure, gate.detected_agents, gate.agent_contention,
    )
}

/// Agent RSS companion line (gate snapshot field).
pub fn format_gate_rss_tray_line(gate: &GateStatusSnapshot) -> String {
    format!("Agent RSS: {}", format_bytes_compact(gate.agent_total_rss_bytes))
}

/// Host load / FD / RSS line for tray menus / status bars.
pub fn format_host_watch_tray_line(host: &HostResourceWatchJson) -> String {
    format!(
        "Host load {:.2} · FDs {} · RSS {}",
        host.load_1m,
        host.fd_count,
        format_bytes_compact(host.mem_rss_bytes),
    )
}

/// Host network RX/TX line for tray menus / status bars.
pub fn format_host_net_tray_line(host: &HostResourceWatchJson) -> String {
    format!(
        "Net RX {} · TX {}",
        format_bytes_compact(host.net_rx_bytes),
        format_bytes_compact(host.net_tx_bytes),
    )
}

/// Ordered tray menu / tooltip lines: gate → host_watch (AC-007.56).
pub fn format_operator_tray_lines(
    gate: &GateStatusSnapshot,
    host: &HostResourceWatchJson,
) -> Vec<String> {
    vec![
        format_gate_tray_line(gate),
        format_gate_rss_tray_line(gate),
        format_host_watch_tray_line(host),
        format_host_net_tray_line(host),
    ]
}

/// Single-line operator summary for compact status bars (Windows header row 1).
pub fn format_operator_status_summary(
    gate: &GateStatusSnapshot,
    host: &HostResourceWatchJson,
) -> String {
    format!(
        "{} | {} | {}",
        format_gate_tray_line(gate),
        format_gate_rss_tray_line(gate),
        format_host_watch_tray_line(host),
    )
}

/// Primary tooltip summary line for Linux SNI (AC-007.61).
pub fn format_tray_tooltip_summary_line(
    visual: &TrayGateVisual,
    health: &HealthSnapshot,
) -> String {
    format!(
        "{} managed · {} / {} MB · {}",
        health.managed_processes, health.used_memory_mb, health.total_memory_mb, visual.badge_label,
    )
}

/// Offline tooltip summary when monitoring.report is unavailable (AC-007.61).
pub fn format_tray_tooltip_offline_line(visual: &TrayGateVisual) -> String {
    format!("ShareCLI · {} · daemon not reachable", visual.badge_label)
}

/// Primary SNI menu header line (AC-007.63).
pub fn format_tray_menu_header_line(visual: &TrayGateVisual, health: &HealthSnapshot) -> String {
    format!(
        "{} · {} process(es) · {} / {} MB",
        visual.badge_label, health.managed_processes, health.used_memory_mb, health.total_memory_mb,
    )
}

/// Offline SNI menu header when monitoring.report is unavailable (AC-007.63).
pub fn format_tray_menu_header_offline_line(visual: &TrayGateVisual) -> String {
    format!("{} · Daemon offline", visual.badge_label)
}

/// Pool capacity line from `pool.status` IPC (AC-007.69).
pub fn format_pool_tray_line(pool: &PoolSnapshot) -> String {
    format!(
        "Pool node {}/{} idle · bun {}/{} idle · max {} · {}",
        pool.node_total,
        pool.node_idle,
        pool.bun_total,
        pool.bun_idle,
        pool.max_per_type,
        if pool.healthy { "healthy" } else { "degraded" },
    )
}

/// Proc-scan summary line from `status.snapshot` IPC (AC-007.69).
pub fn format_status_snapshot_tray_line(status: &StatusSnapshot) -> String {
    format!(
        "Proc scan {} · watched {} · {} managed · {} agent row(s)",
        status.scanned,
        status.watched,
        status.total_processes,
        status.agents.len(),
    )
}

/// Supplementary pool + status operator lines (AC-007.69).
pub fn format_pool_status_operator_lines(
    pool: &PoolSnapshot,
    status: &StatusSnapshot,
) -> Vec<String> {
    vec![format_pool_tray_line(pool), format_status_snapshot_tray_line(status)]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ipc::{GateStatusSnapshot, HostResourceWatchJson};

    fn sample_gate() -> GateStatusSnapshot {
        GateStatusSnapshot {
            thermal_pressure: "YELLOW".into(),
            detected_agents: 1,
            agent_total_rss_bytes: 512,
            agent_contention: "WARN".into(),
            gate_decision: "THROTTLE".into(),
        }
    }

    fn sample_host() -> HostResourceWatchJson {
        HostResourceWatchJson {
            fd_count: 10,
            net_rx_bytes: 100,
            net_tx_bytes: 200,
            mem_rss_bytes: 4096,
            load_1m: 1.25,
        }
    }

    #[test]
    fn format_gate_tray_line_includes_key_fields() {
        let line = format_gate_tray_line(&sample_gate());
        assert!(line.contains("THROTTLE"));
        assert!(line.contains("YELLOW"));
        assert!(line.contains("agents 1"));
        assert!(line.contains("WARN"));
    }

    #[test]
    fn format_host_watch_tray_lines_include_key_fields() {
        let host = sample_host();
        let watch = format_host_watch_tray_line(&host);
        assert!(watch.contains("load 1.25"));
        assert!(watch.contains("FDs 10"));
        assert!(watch.contains("RSS 4.0 KB"));

        let net = format_host_net_tray_line(&host);
        assert!(net.contains("RX 100 B"));
        assert!(net.contains("TX 200 B"));
    }

    #[test]
    fn format_operator_tray_lines_gate_before_host_watch() {
        let lines = format_operator_tray_lines(&sample_gate(), &sample_host());
        assert_eq!(lines.len(), 4);
        assert!(lines[0].starts_with("Gate ["));
        assert!(lines[1].starts_with("Agent RSS:"));
        assert!(lines[2].starts_with("Host load"));
        assert!(lines[3].starts_with("Net RX"));
    }

    #[test]
    fn resolve_tray_gate_visual_admit_green_normal() {
        let v = resolve_tray_gate_visual("GREEN", "ADMIT", true);
        assert_eq!(v.severity, TrayGateSeverity::Normal);
        assert_eq!(v.decision_class, "gate-admit");
        assert_eq!(v.thermal_class, "");
        assert_eq!(v.color_hex, "#3fb950");
        assert_eq!(v.badge_label, "Normal");
        assert_eq!(v.linux_icon_name, "utilities-system-monitor");
    }

    #[test]
    fn resolve_tray_gate_visual_deny_red_critical() {
        let v = resolve_tray_gate_visual("RED", "DENY", true);
        assert_eq!(v.severity, TrayGateSeverity::Critical);
        assert_eq!(v.decision_class, "gate-deny");
        assert_eq!(v.thermal_class, "critical");
        assert_eq!(v.badge_label, "Critical");
        assert_eq!(v.linux_icon_name, "dialog-error");
    }

    #[test]
    fn resolve_tray_gate_visual_throttle_warning() {
        let v = resolve_tray_gate_visual("GREEN", "THROTTLE", true);
        assert_eq!(v.severity, TrayGateSeverity::Warning);
        assert_eq!(v.decision_class, "gate-unavailable");
        assert_eq!(v.badge_label, "Warning");
    }

    #[test]
    fn resolve_tray_gate_visual_offline_when_disconnected() {
        let v = resolve_tray_gate_visual("GREEN", "ADMIT", false);
        assert_eq!(v.severity, TrayGateSeverity::Offline);
        assert_eq!(v.badge_label, "Offline");
    }

    #[test]
    fn format_tray_menu_header_line_prefixes_badge() {
        let visual = resolve_tray_gate_visual("GREEN", "ADMIT", true);
        let health = crate::ipc::HealthSnapshot {
            managed_processes: 2,
            used_memory_mb: 512,
            total_memory_mb: 8192,
            healthy: true,
            gate: sample_gate(),
            host_watch: sample_host(),
        };
        let line = format_tray_menu_header_line(&visual, &health);
        assert_eq!(line, "Normal · 2 process(es) · 512 / 8192 MB");
    }

    #[test]
    fn format_tray_menu_header_offline_line_prefixes_badge() {
        let visual = resolve_tray_gate_visual("GREEN", "ADMIT", false);
        let line = format_tray_menu_header_offline_line(&visual);
        assert_eq!(line, "Offline · Daemon offline");
    }
}
